internal/controllers/topology/cluster/cluster_controller.go

```diff
@@ -181,6 +181,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 	if err := r.Client.Get(ctx, key, clusterClass); err != nil {
 		return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve ClusterClass %s", cluster.Spec.Topology.Class)
 	}
+
+	// If the ClusterClass `metadata.Generation` doesn't match the `status.ObservedGeneration` requeue as the ClusterClass
+	// is not up to date.
+	if clusterClass.GetGeneration() != clusterClass.Status.ObservedGeneration {
+		log.V(2).Info("ClusterClass not up to date. Requeuing", "ClusterClass", clusterClass.Name)
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
Didn't want to return an error here as most of the time this isn't an error state, and shouldn't be considered a Cluster Topology error state. Instead this should be bubbled up on the ClusterClass, and in the logs of the ClusterClass reconciler.

Reply:
With this current logic if the cluster controller returns here because the ClusterClass is out of state I believe the `TopologyReconciled` condition will be set to true. Not sure if that is right. A better signal would be to actually set the `TopologyReconciled` condition to false and set the message to something like `Waiting for ClusterClass to be reconciled`. WDYT?

Reply:
Hmm - I think maybe this should be a follow with conditions and communication. If this fails there's no error, as most of the time this happens it's going to be a minor syncing issue. So when it returns here it won't impact the previous topology reconciled condition.

You are right about the validating and defaulting though - if they fail they return early before the condition is changed. I'll add a topic for that on the umbrella issue, but I don't know if we can solve it all as part of this PR.

Reply:
> With this current logic if the cluster controller returns here because the ClusterClass is out of state I believe the TopologyReconciled condition will be set to true. 

I think the status won't be updated at all as we don't even get to the defer.

> If this fails there's no error, as most of the time this happens it's going to be a minor syncing issue. So when it returns here it won't impact the previous topology reconciled condition.

If the variable discovery is broken for some reason, the cluster will stay in that state indefinitely.

I think TopologyReconciled true is sending the wrong message in this case (if it was set before)

I wonder if we should move this check + defaulting + varlidation into r.reconcile. Does it have to be here already?

I wonder if we should just return an error if the generation doesn't match. The alternative is that we have to detect how long the clusterclass is already in this state (that observedGeneration != metadata.generation).

Reply:
>check + defaulting + varlidation into r.reconcile
Good idea I think. I've made the generation check return an error, but I think it could cause a bunch of extra unnecessary reconciliation. Fine for right now though and we can follow up later on.